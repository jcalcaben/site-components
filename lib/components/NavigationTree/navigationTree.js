import React from "react";

import defaultStyles from "./navigationTree.module.css";

const NavigationTree = () => {
  return <div className={defaultStyles.root}>Navigation</div>;
};

export default NavigationTree;
