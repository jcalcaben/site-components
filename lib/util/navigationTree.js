/**
 * A node representing an entry in a navigation tree
 *
 * @typedef NavigationTreeNode
 *
 * @property {string} label Label to render
 * @property {string} url The url to navigate to when clicked
 * @property {NavigationTreeNode[]} pages Child pages of the node
 */

/**
 *
 * @param {NavigationTreeNode[]} tree An array of tree-like objects
 * @param {NavigationTreeNode} target Target navigation tree node
 *
 * @returns {Object[]} An array of objects containing the url and label
 * for the target and its parents in the tree
 */
export const getPath = (tree, target) => {
  let result = [];
  if (tree) {
    tree.forEach((element) => {
      const { label, url, pages } = element;
      if (url === target) {
        result.push({ label, url });
      } else if (pages) {
        const childrenResults = getPath(pages, target);
        if (childrenResults.length > 0) {
          result = result.concat([{ label, url }], childrenResults);
        }
      }
    });
  }
  return result;
};

export default {};
