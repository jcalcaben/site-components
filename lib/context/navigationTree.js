import React, { useState, createContext, useContext } from "react";
import PropTypes from "prop-types";

import { getPath } from "../util/navigationTree";

const NavigationTreeContext = createContext();

const NavigationTreeContextProvider = (props) => {
  const { children, data } = props;

  const [openSections, setOpenSections] = useState([]);

  const state = {
    data,
    openSections,
  };

  const openSection = (id) => {
    const treePath = getPath(data.pages, id);

    // Prevents duplicate entries
    const newSections = [];

    treePath.forEach((path) => {
      const openSectionsEntry = openSections.find((section) => {
        return section.url === path.url && section.label === path.label;
      });

      if (!openSectionsEntry) {
        newSections.push(path);
      }
    });

    setOpenSections((current) => {
      return [...newSections, ...current];
    });
  };

  const api = {
    openSection,
  };

  const value = [state, api];

  return (
    <NavigationTreeContext.Provider value={value}>
      {children}
    </NavigationTreeContext.Provider>
  );
};

NavigationTreeContextProvider.propTypes = {
  children: PropTypes.element,
  data: PropTypes.object,
};

export default NavigationTreeContextProvider;

export const useNavigationTreeContext = () => useContext(NavigationTreeContext);
