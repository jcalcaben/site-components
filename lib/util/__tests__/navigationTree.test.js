import { getPath } from "../navigationTree";

const data = [
  {
    label: "Section-1",
    url: "section-1",
  },
  {
    label: "Section-2",
    url: "section-2",
    pages: [
      {
        label: "Section-2-1",
        url: "section-2-1",
      },
      {
        label: "Section-2-2",
        url: "section-2-2",
      },
      {
        label: "Section-2-3",
        pages: [
          {
            label: "Section-2-3-1",
            url: "section-2-3-1",
          },
          {
            label: "Section-2-3-2",
            url: "section-2-3-2",
          },
        ],
      },
    ],
  },
  {
    label: "Section-3",
    pages: [
      {
        label: "Section-3-1",
        url: "section-3-1",
      },
    ],
  },
];

test("entry not found", () => {
  const result = getPath(data, "foo");

  expect(result).toEqual([]);
});

test("top level entry", () => {
  const result = getPath(data, "section-1");

  expect(result).toMatchSnapshot();
});

test("second level entry", () => {
  const result = getPath(data, "section-2-2");

  expect(result).toMatchSnapshot();
});

test("third level entry", () => {
  const result = getPath(data, "section-2-3-2");

  expect(result).toMatchSnapshot();
});

test("no url in parent", () => {
  const result = getPath(data, "section-3-1");

  expect(result).toMatchSnapshot();
});
