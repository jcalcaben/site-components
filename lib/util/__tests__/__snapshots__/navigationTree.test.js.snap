// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`no url in parent 1`] = `
Array [
  Object {
    "label": "Section-3",
    "url": undefined,
  },
  Object {
    "label": "Section-3-1",
    "url": "section-3-1",
  },
]
`;

exports[`second level entry 1`] = `
Array [
  Object {
    "label": "Section-2",
    "url": "section-2",
  },
  Object {
    "label": "Section-2-2",
    "url": "section-2-2",
  },
]
`;

exports[`third level entry 1`] = `
Array [
  Object {
    "label": "Section-2",
    "url": "section-2",
  },
  Object {
    "label": "Section-2-3",
    "url": undefined,
  },
  Object {
    "label": "Section-2-3-2",
    "url": "section-2-3-2",
  },
]
`;

exports[`top level entry 1`] = `
Array [
  Object {
    "label": "Section-1",
    "url": "section-1",
  },
]
`;
